Add fritz set guest wifi password service

sti0 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a new service to the `fritz` integration to allow setting a password for the guest wifi. With this service it's possible to add automations to cycle the password for the guest network and make it more secure. It's possible to set a password by own or it's also possible to use the password creation of the underlying `fritzconnection` library. If no parameter is passed to the service the default password will be created with 32 characters.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20915

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Add a new service set_guest_wifi_password to the fritz integration.

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#62892).

Once you do that we will be able to review and accept this pull request.

Thanks!

sti0 commented:
Sry everybody...accidentally added you through rebase process 🤕

chemelli74 commented on homeassistant/components/fritz/common.py:
```diff
@@ -264,6 +271,19 @@ def signal_device_update(self) -> str:
+        """Set the guest networks wifi password."""
+        try:
+            self.fritz_guest_wifi.set_password(password, length)
+        except FritzArgumentStringToShortError as ex:
+            raise HomeAssistantError("Password must be at least 8 characters.") from ex
+        except FritzArgumentStringToLongError as ex:
+            raise HomeAssistantError(
+                "Password must have a maximum length of 63 characters."
+            ) from ex
```
```suggestion
        except (FritzArgumentStringToShortError,FritzArgumentStringToLongError) as ex:
            raise HomeAssistantError("Password must be between 8 and 63 characters length.") from ex
```

chemelli74 commented on homeassistant/components/fritz/services.yaml:
```diff
@@ -35,3 +35,31 @@ cleanup:
+      required: false
+      selector:
+        text:
+    length:
+      name: Password length
+      description: Length of the new password. The password will be auto-generated, if no password is set.
+      required: false
+      default: 32
+      selector:
+        number:
+          min: 8
+          max: 63
```
Why asking for the length to the end user ?
I would say, use your own password or a default (32bytes long) pwd will be used.
That's it

sti0 replied:
Why not? Its no big deal. I one wants a shorter or longer password but like to rotate it for security reason he could use the built-in mechanism and don't needs to worry about how to generate the password randomly.

chemelli74 replied:
I'm not a fan of too many options, honestly.
Will ask for a second opinion on that.

Simone

rytilahti replied:
Separation of concerns. A service that changes a password to `x` differs from a one that generates a random, time-limited password. For the latter, would it not be necessary to have a way to receive the generated output (which is not possible using service calls)?

I have no idea about the internals of this integration, I just stumbled upon it as I got notified by github, so take my advise with a grain of salt.

sti0 replied:
It's my plan to integrate a camera entity for QR code next. So you could run the password change service and the QR code will be updated. This is my personal use-case.

chemelli74 replied:
@mib1185 what you think about that ?

mib1185 replied:
I'm fine with this optional parameter, but would remove the default of 32, since the fritzconnection lib does already have an own [default value](https://github.com/kbr/fritzconnection/blob/0fb1e2913c4c90428be11676e52e5548cee5bb8b/fritzconnection/lib/fritzwlan.py#L28) implemented for that.
For your @sti0 use case, you can for sure set the `length` to what ever you need.

sti0 replied:
Removed the default value. Now if no value is set the default value of `fritzconnection` will be used (length = 12). I will change that in the docs, too.

chemelli74 commented:
You should take care of the conflict as well

sti0 commented:
> You should take care of the conflict as well

Doing it right now ;)

chemelli74 commented on homeassistant/components/fritz/common.py:
```diff
@@ -264,6 +271,15 @@ def signal_device_update(self) -> str:
         return f"{DOMAIN}-device-update-{self._unique_id}"
 
+    def _set_guest_wifi_password(
+        self, password: str = "", length: int = DEFAULT_GUEST_WIFI_PW_LENGTH
```
```suggestion
        self, password: str, length: int
```

You already specify the params in the call, no need to load a default

Commit:
Remove unnecessary params defaults

mib1185 commented on homeassistant/components/fritz/common.py:
```diff
@@ -257,6 +264,15 @@ def signal_device_update(self) -> str:
+        """Set the guest networks wifi password."""
+        try:
+            self.fritz_guest_wifi.set_password(password, length)
+        except (FritzArgumentStringToShortError, FritzArgumentStringToLongError) as ex:
+            raise HomeAssistantError(
+                "Password must be between 8 and 63 characters length."
+            ) from ex
```
you can add a validation schema while registering the service (see [`hass.services.async_register`](https://dev-docs.home-assistant.io/en/master/api/core.html?highlight=services.async_register#homeassistant.core.ServiceRegistry.async_register)), so invalid inputs are handled proper by HA itself

sti0 replied:
I've tried this but can't get it to work. Maybe you could guide me.

I created this schema  in `conf.py`:
```
SERVICE_SCHEMA_GUEST_WIFI_PW_LENGTH = vol.Schema(
    {vol.Optional("length"): vol.Range(min=8, max=63)}
)
```

But I don't really understand how to use it and at which part. Must the service call be in a own function instead of `services_fritzbox`? And where to call `hass.services.async_register` than?

mib1185 replied:
the schema must contain all parameters and needs to be added during `hass.services.async_register()` - this is done here:
https://github.com/home-assistant/core/blob/09408234a6e07a0e6d6fb709f750dae9f9f69d65/homeassistant/components/fritz/services.py#L55

So the [`SERVICE_LIST`](https://github.com/home-assistant/core/blob/09408234a6e07a0e6d6fb709f750dae9f9f69d65/homeassistant/components/fritz/services.py#L21) needs to be extended, so that a schema can optionally be included.

sti0 replied:
Thanks for the hint. Added a service schema now :)

mib1185 commented on homeassistant/components/fritz/common.py:
```diff
@@ -257,6 +264,15 @@ def signal_device_update(self) -> str:
         """Event specific per FRITZ!Box entry to signal updates in devices."""
         return f"{DOMAIN}-device-update-{self._unique_id}"
 
+    def _set_guest_wifi_password(self, password: str, length: int) -> None:
```
btw. this method seems not to be used any where in the code? 🤔

sti0 replied:
My fault, implemented it in the most recent change. But if we could do the schema validation, it's not necessary and can be removed.

Commit:
Remove default password length

mib1185 commented:
since the AvmWrapper is now introduced _(see #63682 #64133 #64667_), please rebase to latest `dev` branch and adopt your code properly - at least:
- move `_set_guest_wifi_password` to the new wrapper class
- make it sync and async available

cc: @chemelli74

Commit:
Add service schema, cleanup code

sti0 commented:
I removed `_set_guest_wifi_password` and added a `async_trigger_set_guest_password` function to `FritzBoxTools` is this ok? Or should I move all new logic to `AvmWrapper`. I don't know what your plan with `FritzboxTools` and `AvmWrapper` is.

chemelli74 commented:
@sti0, please rebase. Should take care of conflict and failing tests.

Simone

Commit:
Fix min password length in services.yaml

sti0 commented:
@chemelli74 done that. Looking forward for your review. Thanks.

mib1185 commented on homeassistant/components/fritz/services.py:
```diff
@@ -13,18 +13,25 @@
 
 
-SERVICE_LIST = [SERVICE_CLEANUP, SERVICE_REBOOT, SERVICE_RECONNECT]
+SERVICE_LIST = [
```
```suggestion
SERVICE_LIST: list[tuple[str, vol.Schema | None]] = [
```
would add some typing

mib1185 commented on homeassistant/components/fritz/const.py:
```diff
@@ -49,6 +50,15 @@ class MeshRoles(StrEnum):
 SERVICE_CLEANUP = "cleanup"
+SERVICE_SET_GUEST_WIFI_PW = "set_guest_wifi_password"
+
+SERVICE_SCHEMA_SET_GUEST_WIFI_PW = vol.Schema(
+    {
+        vol.Required("device_id"): str,
+        vol.Optional("password"): vol.Length(min=8, max=63),
+        vol.Optional("length"): vol.Range(min=8, max=63),
+    }
+)
```
The schema definition should be moved into the `services.py` since it it is only used there

mib1185 commented on homeassistant/components/fritz/common.py:
```diff
@@ -512,6 +524,13 @@ async def service_fritzbox(
+            if service_call.service == SERVICE_SET_GUEST_WIFI_PW:
+                await self.async_trigger_set_guest_password(
+                    service_call.data.get("password"),
+                    service_call.data.get("length"),
```
use `DEFAULT_PASSWORD_LENGTH` from `fritzconnection.lib.fritzwlan` as default

```suggestion
                    service_call.data.get("length", DEFAULT_PASSWORD_LENGTH),
```
the `length` parameter in `fritz_guest_wifi.set_password` needs to be an `int`

sti0 replied:
But if `None` is passed to the `set_password` method it already has the `DEFAULT_PASSWORD_LENGHT` as a default parameter. So this should not be necessary?!

sti0 replied:
https://github.com/kbr/fritzconnection/blob/master/fritzconnection/lib/fritzwlan.py#L252

mib1185 replied:
You explicit set the `length` to the result of `service_call.data.get("length")` ... so if it returns `None` than `length` will be set to None

sti0 replied:
Right...silly me :-/

mib1185 reviewed (changes requested):
Looks pretty good, only some comments

Commit:
Move schema to `services.py`, add typing

Commit:
Add default password length from upstream lib

sti0 commented:
@mib1185 thanks for your suggestions. I think this one is now good to go.

mib1185 commented on homeassistant/components/fritz/common.py:
```diff
@@ -413,6 +417,14 @@ async def async_trigger_reconnect(self) -> None:
         await self.hass.async_add_executor_job(self.connection.reconnect)
 
+    async def async_trigger_set_guest_password(
+        self, password: str | None, length: int | None
```
```suggestion
        self, password: str | None, length: int
```

sorry, did oversee this in the last review 🙄

mib1185 reviewed (commented):
Just one small, than we can go 👍

sti0 committed:
Remove None typing

Co-authored-by: Michael <35783820+mib1185@users.noreply.github.com>

sti0 commented:
Lets go before we find more of those 😆

mib1185 reviewed (approved):
LGTM 👍  thx @sti0

chemelli74 reviewed (approved):
LGTM! Thx @sti0
